docs/architecture/adr-008-event-subscription.md

```diff
@@ -0,0 +1,519 @@
+WebSocket connection to provide subscription functionality (the
+`WebSocketSubscriptionClient` in the entity diagram above).
+
+```rust
+#[async_trait]
+pub trait SubscriptionClient {
+    /// `/subscribe`: subscribe to receive events produced by the given query,
+    /// but specify how many event results can be buffered in the resulting
+    /// subscription.
+    ///
+    /// Specifying a `buf_size` of zero should indicate to the transport that an
+    /// **unbounded** channel should be used.
+    async fn subscribe_with_buf_size(
+        &mut self,
+        query: String,
+        buf_size: usize,
+    ) -> Result<Subscription>;
+
+    /// `/subscribe`: subscribe to receive events produced by the given query.
+    async fn subscribe(&mut self, query: String) -> Result<Subscription> {
+        // Use an unbounded channel by default
+        self.subscribe_with_buf_size(query, 0)
+            .await
+    }
+
+    /// `/unsubscribe`: unsubscribe from receiving events for the given
+    /// subscription.
+    ///
+    /// This terminates the given subscription and consumes it, since it is no
+    /// longer usable.
+    async fn unsubscribe(&mut self, subscription: Subscription) -> Result<()>;
+}
+```
+
+### Client Implementations
+
+We envisage 3 distinct client implementations at this point:
+
+* `HttpClient`, which only implements [`Client`](#client) (over HTTP).
+* `WebSocketSubscriptionClient`, which only implements
+  [`SubscriptionClient`](#subscriptionclient) (over a WebSocket connection).
+* `HttpWebSocketClient`, which implements both [`Client`](#client) (over HTTP)
+  and [`SubscriptionClient`](#subscriptionclient) (over a WebSocket connection).
+
+#### Handle-Driver Concurrency Model
+
+Depending on the underlying transport, a client may need a **transport driver**
+running in an asynchronous context. As in the example of a WebSocket connection,
+the rate at which one interacts with the WebSocket connection may differ to the
+rate at which one interacts with `Event`s being produced by a `Subscription`, so
+they necessarily need to run concurrently in different contexts.
+
+Implementation of such a transport driver is transport-specific. Short-lived
+request/response interactions (such as HTTP) would not require such a transport
+driver, whereas a WebSocket connection would.
+
+In cases where a driver is necessary, the client implementation would have to
+become a **handle** to the driver, facilitating communication with it across
+asynchronous tasks.
+
+### `SubscriptionRouter`
+
+All possible [`SubscriptionClient`](#subscriptionclient) implementations would
+need some form of subscription management and event/result routing. A
+`SubscriptionRouter` is proposed whose interface facilitates:
+
+1. Immediate subscribe/unsubscribe request fulfilment.
+2. Two-stage subscribe/unsubscribe request management (where a
+   subscription/unsubscribe request can first be created in a "pending" state,
```

Comment:
What's the motivation for this?

Reply:
Since the client multiplexes multiple subscriptions on a single WebSocket connection, there are no request/reply mechanics. You send messages and you continuously receive them in any order (i.e. subscription request responses could be mixed in with incoming events related to a different subscription). For example:

1. The client sends a subscription request via the WebSocket connection. This creates a subscription in a `Pending` state, but it's not yet active (since the RPC can still respond with an error). We don't publish events to `Pending` subscriptions.
2. The client may, while waiting for a response, receive events for a different subscription (made earlier than (1)).
3. When a response comes back whose ID matches the request from (1), if it's successful, the subscription is marked as `Active`, and if it failed we cancel and remove it.

Similar dynamics happen when unsubscribing, so unsubscribing is also currently a two-stage process.

Reply:
So when/how can we ever not use the two-stage subscription? ie when can we use the immediate one?